File: solutions/cpp/4881/main.cpp
#include<bits/stdc++.h>

using namespace std;

int next(int N) {
    int ret = 0;
    while(N) {
        int x = N % 10;
        ret += x * x;
        N /= 10;
    }
    return ret;
}

vector<int> get_sequence(int N) {
    int cur = N;
    set<int> st({cur});
    vector<int> ret({cur});
    while(true) {
        int nxt = next(cur);
        if(st.count(nxt)) break;
        st.insert(nxt); ret.push_back(nxt);
        cur = nxt;
    }
    return ret;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    while (true) {
        int A, B; cin >> A >> B;
        if(A == 0 && B == 0) break;
        map<int, int> mp;
        int answer = 0;
        bool flag = false;
        auto sA = get_sequence(A);
        auto sB = get_sequence(B);
        for(int x : sA) mp[x] = mp.size() + 1;
        int mn = INT_MAX;
        for(int x : sB) {
            ++answer;
            if(mp.find(x) == mp.end()) continue;
            flag = true;
            mn = min(mn, answer + mp[x]);
        }
        if(!flag) answer = 0;
        else answer = mn;
        cout << A << " " << B << " " << answer << '\n';
    }

    return 0;
}

File: solutions/python/4881/main.py
import sys
from typing import List

def input() -> str:
    return sys.stdin.readline().rstrip()


def get_next(N: int) -> int:
    ret = 0
    while N > 0:
        x = N % 10
        ret += x * x
        N //= 10
    return ret


def get_sequence(N: int) -> List[int]:
    cur = N
    ret, used = [cur], {cur,}
    while True:
        nxt = get_next(cur)
        if nxt in used: break
        used.add(nxt)
        ret.append(nxt)
        cur = nxt
    return ret


def solution(A: int, B: int) -> int:
    answer, mn = 0, 10**9
    flag = False
    seqA, seqB = get_sequence(A), get_sequence(B)
    d = {x:idx for idx, x in enumerate(seqA, 1)}
    for x in seqB:
        answer += 1
        mn = min(mn, answer + d.get(x, 10 ** 9))
    answer = 0 if mn == 10 ** 9 else mn
    return answer


if __name__ == "__main__":
    answer = []
    while True:
        x, y = map(int, input().split())
        if x == 0 and y == 0: break
        answer.append(f"{x} {y} {solution(x, y)}")
    print('\n'.join(answer))